sydent/hs_federation/verifier.py

```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018 New Vector Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+
+from twisted.internet import defer
+from twisted.names.error import DNSNameError
+import twisted.names.client
+from unpaddedbase64 import decode_base64
+import signedjson.sign
+import signedjson.key
+from signedjson.sign import SignatureVerifyException
+
+from sydent.http.httpclient import FederationHttpClient
+
+
+logger = logging.getLogger(__name__)
+
+"""
+Verifies signed json blobs from Matrix Homeservers by finding the
+Homeserver's address, contacting it, requesting its keys and
+verifying that the signature on the json blob matches.
+"""
+class Verifier(object):
+    def __init__(self, sydent):
+        self.sydent = sydent
+
+    @defer.inlineCallbacks
+    def _getEndpointForServer(self, server_name):
+        if ':' in server_name:
+            defer.returnValue(tuple(server_name.rsplit(':', 1)))
+
+        service_name = "_%s._%s.%s" % ('_matrix', '_tcp', server_name)
+
+        default = server_name, 8448
+
+        try:
+            answers, _, _ = yield twisted.names.client.lookupService(service_name)
+        except DNSNameError:
+            defer.returnValue(default)
+
+        for answer in answers:
+            if answer.type != dns.SRV or not answer.payload:
+                continue
+
+            # XXX we just use the first
```

Comment:
hm, wouldn't it be pretty broken to have two SRV records?  Or is this considering some kind of broken attempt at RRing?

Reply:
SRVs have priority and weight fields, so you can have multiple of them and the client balances between them based on those priority / weight fields.